Mathlib/NumberTheory/Ostrowski.lean

```diff
@@ -343,6 +384,133 @@ lemma one_lt_of_not_bounded (notbdd : ¬ ∀ n : ℕ, f n ≤ 1) {n₀ : ℕ} (h
+lemma le_of_mulRingNorm_eq {s t : ℝ} (hfm : f m = m ^ s) (hfn : f n = n ^ t)  : t ≤ s := by
+    have hmn : f n ≤ f m ^ Real.logb m n := mulRingNorm_le_mulRingNorm_pow_log hm hn notbdd
+    rw [← Real.rpow_le_rpow_left_iff (x:=n) (mod_cast hn), ← hfn]
+    apply le_trans hmn
+    rw [hfm, ← Real.rpow_mul (Nat.cast_nonneg m), mul_comm, Real.rpow_mul (Nat.cast_nonneg m),
+      Real.rpow_logb (mod_cast zero_lt_of_lt hm) (mod_cast Nat.ne_of_lt' hm)
+      (mod_cast zero_lt_of_lt hn)]
+
+include hm hn notbdd in
+private lemma symmetric_roles {s t : ℝ} (hfm : f m = m ^ s) (hfn : f n = n ^ t) : s = t :=
+    le_antisymm (le_of_mulRingNorm_eq hn hm notbdd hfn hfm)
+    (le_of_mulRingNorm_eq hm hn notbdd hfm hfn)
+
+/-! ## Archimedean case: end goal -/
+
+include notbdd in
+/-- If `f` is not bounded and not trivial, then it is equivalent to the standard absolute value on
+`ℚ`. -/
+theorem mulRingNorm_equiv_standard_of_unbounded : MulRingNorm.equiv f mulRingNorm_real := by
+  obtain ⟨m, hm⟩ := Classical.exists_not_of_not_forall notbdd
+  have oneltm : 1 < m := by
+    by_contra!
+    apply hm
+    replace this : m = 0 ∨ m = 1 := by omega
+    rcases this with (rfl | rfl)
+    all_goals simp only [CharP.cast_eq_zero, map_zero, zero_le_one, Nat.cast_one, map_one, le_refl]
+  rw [← equiv_on_nat_iff_equiv]
+  set s := Real.logb m (f m) with hs
+  use s⁻¹
+  refine ⟨inv_pos.2 (Real.logb_pos (Nat.one_lt_cast.2 oneltm)
+    (one_lt_of_not_bounded notbdd oneltm)), ?_⟩
+  intro n
+  by_cases h1 : n ≤ 1
+  · by_cases h2 : n = 1
+    · simp only [h2, Nat.cast_one, map_one, one_rpow, mul_ring_norm_eq_abs, abs_one, cast_one]
+    · have : n = 0 := by omega
+      rw [this, hs]
+      simp only [CharP.cast_eq_zero, map_zero]
+      rw [Real.rpow_eq_zero le_rfl]
+      simp only [ne_eq, inv_eq_zero, logb_eq_zero, Nat.cast_eq_zero, Nat.cast_eq_one, map_eq_zero,
+        not_or]
+      push_neg
+      exact ⟨not_eq_zero_of_lt oneltm, Nat.ne_of_lt' oneltm, mod_cast (fun a ↦ a),
+        not_eq_zero_of_lt oneltm, ne_of_not_le hm, by linarith only [apply_nonneg f ↑m]⟩
+  · simp only [mul_ring_norm_eq_abs, abs_cast, Rat.cast_natCast]
+    rw [Real.rpow_inv_eq (apply_nonneg f ↑n) (Nat.cast_nonneg n)
+      (Real.logb_ne_zero_of_pos_of_ne_one (one_lt_cast.mpr oneltm) (by linarith only [hm])
+      (by linarith only [hm]))]
+    simp only [not_le] at h1
+    have hfm : f m = m ^ s := by rw [Real.rpow_logb (mod_cast zero_lt_of_lt oneltm)
+      (mod_cast Nat.ne_of_lt' oneltm) (by linarith only [hm])]
+    have hfn : f n = n ^ (Real.logb n (f n)) := by
+      rw [Real.rpow_logb (mod_cast zero_lt_of_lt h1) (mod_cast Nat.ne_of_lt' h1)
+      (by apply map_pos_of_ne_zero; exact_mod_cast not_eq_zero_of_lt h1)]
+    rwa [← hs, symmetric_roles oneltm h1 notbdd hfm hfn]
+
 end Archimedean
 
+/-- Ostrowski's Theorem -/
+theorem mulRingNorm_equiv_standard_or_padic (f : MulRingNorm ℚ) (hf_nontriv : f ≠ 1) :
```

Comment:
Can you add the theorem to the doc at the beginning of the file?

Reply:
Done!